# Importação dos dados
O primeiro passo a ser feito é analisar e corrigir os dados se assim for necessário.

<div align = "center">
    <img src = "./IMG/Logo_Alura_Voz.png" alt = "Alura Voz"/>
</div>

In [1]:
# Importando o pandas para fazer a análise 
import pandas as pd

# Importando o pacote pessoal para ajudar na análise e limpeza dos dados. 
import suportedata as sd

In [2]:
# algumas configurações de visualização do DataFrame
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

In [3]:
# importando a primeira visualização do DataFrame
clientes = pd.read_json('dados/Telco-Customer-Churn.json')

In [4]:
clientes

customerID Churn  \
0     0002-ORFBO    No   
1     0003-MKNFE    No   
2     0004-TLHLJ   Yes   
3     0011-IGKFF   Yes   
4     0013-EXCHZ   Yes   
...          ...   ...   
7262  9987-LUTYD    No   
7263  9992-RRAMN   Yes   
7264  9992-UJOEL    No   
7265  9993-LHIEB    No   
7266  9995-HOTOH    No   

                                                                                          customer  \
0     {'gender': 'Female', 'SeniorCitizen': 0, 'Partner': 'Yes', 'Dependents': 'Yes', 'tenure': 9}   
1         {'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'No', 'Dependents': 'No', 'tenure': 9}   
2         {'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'No', 'Dependents': 'No', 'tenure': 4}   
3       {'gender': 'Male', 'SeniorCitizen': 1, 'Partner': 'Yes', 'Dependents': 'No', 'tenure': 13}   
4      {'gender': 'Female', 'SeniorCitizen': 1, 'Partner': 'Yes', 'Dependents': 'No', 'tenure': 3}   
...                                                                                            ...   
7262   {'gender': 'Female', 'SeniorCitizen': 0, 'Partner': 'No', 'Dependents': 'No', 'tenure': 13}   
7263    {'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'Yes', 'Dependents': 'No', 'tenure': 22}   
7264      {'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'No', 'Dependents': 'No', 'tenure': 2}   
7265   {'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'Yes', 'Dependents': 'Yes', 'tenure': 67}   
7266   {'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'Yes', 'Dependents': 'Yes', 'tenure': 63}   

                                                            phone  \
0                  {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1                 {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2                  {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3                  {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4                  {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                           ...   
7262               {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263              {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264               {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265               {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No phone service'}   

                                                                                                                                                                               internet  \
0              {'InternetService': 'DSL', 'OnlineSecurity': 'No', 'OnlineBackup': 'Yes', 'DeviceProtection': 'No', 'TechSupport': 'Yes', 'StreamingTV': 'Yes', 'StreamingMovies': 'No'}   
1                {'InternetService': 'DSL', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'No', 'TechSupport': 'No', 'StreamingTV': 'No', 'StreamingMovies': 'Yes'}   
2        {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'Yes', 'TechSupport': 'No', 'StreamingTV': 'No', 'StreamingMovies': 'No'}   
3     {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'Yes', 'DeviceProtection': 'Yes', 'TechSupport': 'No', 'StreamingTV': 'Yes', 'StreamingMovies': 'Yes'}   
4       {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'No', 'TechSupport': 'Yes', 'StreamingTV': 'Yes', 'StreamingMovies': 'No'}   
...                                                                                                                                                                                 ...   
7262            {'InternetService': 'DSL', 'OnlineSecurity': 'Yes', 'OnlineBackup': 'No', 'DeviceProtection': 'No', 'TechSupport': 'Yes', 'StreamingTV': 'No', 'StreamingMovies': 'No'}   
7263     {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'No', 'TechSupport': 'No', 'Strea

Nesse primeiro momento é possível visualizar duas coisas, a primeira é que os dados se encontra agrupado em colunas, para isso é preciso extrair esses dados em um processo chamado de normalização. E a segunda é que o nosso DataFrame possui 7267 linhas e 6 colunas, essas 7267 linhas são exatamente as informações de todos os clientes que foram disponibilizados.

In [5]:
# foram criados alguns DataFrames de apoio
customer = pd.json_normalize(clientes['customer']) 
phone = pd.json_normalize(clientes['phone']) 
internet = pd.json_normalize(clientes['internet']) 
account = pd.json_normalize(clientes['account'])

# esses DataFrames foram colocados em uma lista 
lista = [customer, phone, internet, account]

# dropando as colunas com informações já normalizadas. 
clientes_v01 = clientes.copy()
clientes_v01.drop(columns=['customer', 'phone','internet','account'], inplace=True)

Após o processo de normalização das colunas foram criados alguns DataFrames de apoio. Como vimos anteriormente, estes DataFrames carregam informações dos clientes e para melhor entender quais sãos essas informações, visualizaremos e identificaremos que  informações são essas que cada um desses novos DataFrames possuem. 

# Verificando as informações

Nesta esta de verificação olharemos quais sãos os valores únicos em cada uma das colunas desses novos DataFrames criados.

> A parti daqui chamaremos as colunas dos nossos DataFrames de **variáveis**.

In [6]:
sd.tipo_variaveis(customer)

Valores da variável gender --> ['Female' 'Male'] 

Valores da variável SeniorCitizen --> [0 1] 

Valores da variável Partner --> ['Yes' 'No'] 

Valores da variável Dependents --> ['Yes' 'No'] 

Valores da variável tenure --> [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 11 37 49 66
 67 20 43 59 12 27  2 25 29 14 35 64 39 40  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 51 31 36 17 18 38 42
  0] 



Podemos observar que no DataFrame `customer` as variáveis são as informação dos clientes, tais como gênero, se este é sênior ou não (possui mais de 65 anos), se ele tem parentes e dependentes e também o tempo que ele é cliente em nossa empresa, dados em meses. 

In [7]:
sd.tipo_variaveis(phone)

Valores da variável PhoneService --> ['Yes' 'No'] 

Valores da variável MultipleLines --> ['No' 'Yes' 'No phone service'] 



Já o DataFrame `phone` que as variáveis são informações sobre o serviço de telefone fornecido pela empresa, sendo assim o usuário pode ou não fazer usufruto desse serviço e possuir ou não varias linhas telefônicas.

In [8]:
sd.tipo_variaveis(internet)

Valores da variável InternetService --> ['DSL' 'Fiber optic' 'No'] 

Valores da variável OnlineSecurity --> ['No' 'Yes' 'No internet service'] 

Valores da variável OnlineBackup --> ['Yes' 'No' 'No internet service'] 

Valores da variável DeviceProtection --> ['No' 'Yes' 'No internet service'] 

Valores da variável TechSupport --> ['Yes' 'No' 'No internet service'] 

Valores da variável StreamingTV --> ['Yes' 'No' 'No internet service'] 

Valores da variável StreamingMovies --> ['No' 'Yes' 'No internet service'] 



O DataFrame `internet`, assim como o DataFrame anterior, trans informação do tipo de serviço de internet, tais como o modelo de internet que o cliente possui e se possui, os serviços de Seguro Online, Backup Online, Serviço de Proteção do Aparelho, Suporte Técnico e os serviços de Streaming de TV e de Filmes.

In [9]:
sd.tipo_variaveis(account)

Valores da variável Contract --> ['One year' 'Month-to-month' 'Two year'] 

Valores da variável PaperlessBilling --> ['Yes' 'No'] 

Valores da variável PaymentMethod --> ['Mailed check' 'Electronic check' 'Credit card (automatic)'
 'Bank transfer (automatic)'] 

Valores da variável Charges.Monthly --> [65.6  59.9  73.9  ... 91.75 68.8  67.85] 

Valores da variável Charges.Total --> ['593.3' '542.4' '280.85' ... '742.9' '4627.65' '3707.6'] 



Já o DataFrame `account`é o que traz as informações de pagamento do cliente como o Tipo de contrato, se o cliente recebe a fatura sem papel, os métodos de pagamento, e traz também o valor gasto mensal e o valor gasto total pelo cliente. 

# Juntando os DataFrames

### DataFrame Original

In [10]:
clientes_v01.head()

customerID Churn
0  0002-ORFBO    No
1  0003-MKNFE    No
2  0004-TLHLJ   Yes
3  0011-IGKFF   Yes
4  0013-EXCHZ   Yes

Como é possível ver, cada um destes DataFrames secundários trazem informações diferentes sobre os mesmos usuário, sendo assim, juntá-los ao DataFrame original no qual somete ficaram as informações de **Churn** e os IDs dos clientes é fundamental para podermos fazer melhores analises de nossos dados.

### Novo DataFrame

In [11]:
clientes_v02 = pd.concat([clientes_v01, customer, phone, internet, account], axis=1)
clientes_v02.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService OnlineSecurity OnlineBackup  \
0          Yes            No             DSL             No          Yes   
1          Yes           Yes             DSL             No           No   
2          Yes            No     Fiber optic             No           No   
3          Yes            No     Fiber optic             No          Yes   
4          Yes            No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod  Charges.Monthly Charges.Total  
0              Yes      Mailed check             65.6         593.3  
1               No      Mailed check             59.9         542.4  
2              Yes  Electronic check             73.9        280.85  
3              Yes  Electronic check             98.0       1237.85  
4              Yes      Mailed check             83.9         267.4

# Tradução das Variáveis

Antes de dar continuidade, farei uma tradução das variáveis(Colunas), de modo a facilitar nossa compreensão e também para evitar possíveis problemas na seleção das variáveis.

In [12]:
variaveis = {
    'gender' : 'genero',
    'SeniorCitizen' : 'senior',
    'Partner' : 'conjuge',
    'Dependents' : 'dependentes',
    'tenure' : 'permanencia',
    'PhoneService' : 'ServicoTelefonico',
    'MultipleLines' : 'MultiplasLinhas',
    'InternetService': 'ServicoInternet',
    'OnlineSecurity' : 'SeguroOnline',
    'DeviceProtection' : 'ProtecaoDoDispositivo',
    'TechSupport' : 'SuporteTecnico',
    'StreamingMovies' : 'StreamingFilmes',
    'Contract' : 'contrato',
    'PaperlessBilling' : 'FaturaOnline',
    'PaymentMethod' : 'MetodoDePagamento',
    'Charges.Monthly' : 'GastoMensal',
    'Charges.Total' : 'GastoTotal'
}
valores = {
      'Female' : 'Mulher',
    'Male' : 'Homem',
    'Yes' : 'Sim',
    'No' : 'Não',
    'No phone service' : 'SemTelefone',
    'No internet service' : 'SemInternet',
    'Fiber optic' :'FibraOptica',
    'One year' : 'UmAno',
    'Two year' : 'DoisAnos',
    'Month-to-month' : 'Mensal',
    'Mailed check' : 'EnvioDeCheque',
    'Electronic check' : 'ChequeEletronico',
    'Credit card (automatic)' : 'CartaoCredito',
    'Bank transfer (automatic)' : 'TransferenciaBancaria'    
}


clientes_v03 = clientes_v02.rename(columns = variaveis)
clientes_v03.replace(to_replace = valores, inplace = True)
clientes_v03.head()

customerID Churn  genero  senior conjuge dependentes  permanencia  \
0  0002-ORFBO   Não  Mulher       0     Sim         Sim            9   
1  0003-MKNFE   Não   Homem       0     Não         Não            9   
2  0004-TLHLJ   Sim   Homem       0     Não         Não            4   
3  0011-IGKFF   Sim   Homem       1     Sim         Não           13   
4  0013-EXCHZ   Sim  Mulher       1     Sim         Não            3   

  ServicoTelefonico MultiplasLinhas ServicoInternet SeguroOnline OnlineBackup  \
0               Sim             Não             DSL          Não          Sim   
1               Sim             Sim             DSL          Não          Não   
2               Sim             Não     FibraOptica          Não          Não   
3               Sim             Não     FibraOptica          Não          Sim   
4               Sim             Não     FibraOptica          Não          Não   

  ProtecaoDoDispositivo SuporteTecnico StreamingTV StreamingFilmes contrato  \
0                   Não            Sim         Sim             Não    UmAno   
1                   Não            Não         Não             Sim   Mensal   
2                   Sim            Não         Não             Não   Mensal   
3                   Sim            Não         Sim             Sim   Mensal   
4                   Não            Sim         Sim             Não   Mensal   

  FaturaOnline MetodoDePagamento  GastoMensal GastoTotal  
0          Sim     EnvioDeCheque         65.6      593.3  
1          Não     EnvioDeCheque         59.9      542.4  
2          Sim  ChequeEletronico         73.9     280.85  
3          Sim  ChequeEletronico         98.0    1237.85  
4          Sim     EnvioDeCheque         83.9      267.4

# Verificando as Variáveis

Agora que já temos o nosso novo DataFrame para poder trabalhar, precisamos fazer a verificação dos valores e das inconsistências de cada variável, para que nossa análise seja a mais confiável possível.

In [13]:
clientes_v02.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


Em uma primeira análise podemos verificar que não existem elementos faltando em nenhuma das variáveis, porém falta olhar ainda as duas variáveis que ficaram no nosso DataFrame original, a variável `Churn` e a variável `customerID`. E existem outras inconsistências no nosso DataFrame como, por exemplo, a variável `Charges.Total` esta identificada como `object`, sendo que já havíamos visto anteriormente que ela se travada do valor acumulado gasto por um cliente, ou seja, ela deveria esta como `float` (número decimal).

## Verificando a variável `Churn`

In [14]:
print('Variável Churn:')
print(clientes_v02['Churn'].unique())

Variável Churn:
['No' 'Yes' '']


In [15]:
# query para verificar os elementos incorretos 
clientes_v03.query('Churn == ""')

customerID Churn  genero  senior conjuge dependentes  permanencia  \
30    0047-ZHDTW        Mulher       0     Não         Não           11   
75    0120-YZLQA         Homem       0     Não         Não           71   
96    0154-QYHJU         Homem       0     Não         Não           29   
98    0162-RZGMZ        Mulher       1     Não         Não            5   
175   0274-VVQOQ         Homem       1     Sim         Não           65   
...          ...   ...     ...     ...     ...         ...          ...   
7158  9840-GSRFX        Mulher       0     Não         Não           14   
7180  9872-RZQQB        Mulher       0     Sim         Não           49   
7211  9920-GNDMB         Homem       0     Não         Não            9   
7239  9955-RVWSC        Mulher       0     Sim         Sim           67   
7247  9966-VYRTZ        Mulher       0     Sim         Sim           31   

     ServicoTelefonico MultiplasLinhas ServicoInternet SeguroOnline  \
30                 Sim             Sim     FibraOptica          Sim   
75                 Sim             Não             Não  SemInternet   
96                 Sim             Não             DSL          Sim   
98                 Sim             Não             DSL          Sim   
175                Sim             Sim     FibraOptica          Não   
...                ...             ...             ...          ...   
7158               Sim             Sim             DSL          Não   
7180               Não     SemTelefone             DSL          Sim   
7211               Sim             Sim     FibraOptica          Não   
7239               Sim             Não             Não  SemInternet   
7247               Sim             Não             Não  SemInternet   

     OnlineBackup ProtecaoDoDispositivo SuporteTecnico  StreamingTV  \
30            Não                   Não            Não          Não   
75    SemInternet           SemInternet    SemInternet  SemInternet   
96            Sim                   Não            Sim          Não   
98            Sim                   Não            Sim          Não   
175           Sim                   Sim            Não          Sim   
...           ...                   ...            ...          ...   
7158          Sim                   Não            Não          Não   
7180          Não                   Não            Não          Sim   
7211          Não                   Não            Não          Não   
7239  SemInternet           SemInternet    SemInternet  SemInternet   
7247  SemInternet           SemInternet    SemInternet  SemInternet   

     StreamingFilmes  contrato FaturaOnline      MetodoDePagamento  \
30               Não    Mensal          Sim  TransferenciaBancaria   
75       SemInternet  DoisAnos          Sim          CartaoCredito   
96               Não     UmAno          Sim       ChequeEletronico   
98               Não    Mensal          Não          CartaoCredito   
175              Sim     UmAno          Sim  TransferenciaBancaria   
...              ...       ...          ...                    ...   
7158             Não     UmAno          Sim          EnvioDeCheque   
7180             Não    Mensal          Não  TransferenciaBancaria   
7211             Não    Mensal          Sim       ChequeEletronico   
7239     SemInternet  DoisAnos          Sim  TransferenciaBancaria   
7247     SemInternet    Mensal          Sim          EnvioDeCheque   

      GastoMensal GastoTotal  
30          79.00      929.3  
75          19.90     1355.1  
96          58.75     1696.2  
98          59.90     287.85  
175        103.15    6792.45  
...           ...        ...  
7158        54.25      773.2  
7180        40.65    2070.75  
7211        76.25     684.85  
7239        19.25     1372.9  
7247        19.55     658.95  

[224 rows x 21 columns]

Mesmo que em uma primeira análise não vimos elementos faltando em nenhuma das variáveis, na variável `Churn` existem 224 elementos “em branco”, são elementos que não foram preenchidos adequadamente. Neste ponto é importa analisar se esta falta de informação ira impactar em nossa análise. Tendo em vista que nosso objetivo é justamente criar um modelo para a prevenção dos cancelamentos, estes dados que estão preenchidos inadequadamente não nos beneficia em nada, podendo até mesmo nos prejudicar, sendo assim, optarei por removê-los do nosso DataFrame.

In [16]:
# excluindo os elementos incorretos
clientes_v04 = clientes_v03[~(clientes_v02['Churn'] == '')]

# refazendo a query para verifica se os elementos errados ainda estão no DataFrame 
churn_nao_preenchidos = clientes_v04.query('Churn == ""').shape[0]

print(f'Total de não preenchidos na coluna Churn:{churn_nao_preenchidos}')

Total de não preenchidos na coluna Churn:0


## Variável 'customerID'

In [17]:
clientes_v04['customerID'].value_counts()

0002-ORFBO    1
6616-AALSR    1
6625-UTXEW    1
6625-IUTTT    1
6625-FLENO    1
             ..
3352-RICWQ    1
3352-ALMCK    1
3351-NQLDI    1
3351-NGXYI    1
9995-HOTOH    1
Name: customerID, Length: 7043, dtype: int64

Outro ponto a se verificar do DataFrame original é a coluna `customerID`. Essa coluna é usada na identificação de cada cliente, sendo assim cada cliente possui um ID único, e em nosso estudo esta informação não ira nos acrescentar valor, sendo assim ela sera excluída de nossa DataFrame.

In [18]:
clientes_v05 = clientes_v03.drop(columns=['customerID'], axis= 1)
clientes_v05.head()

Churn  genero  senior conjuge dependentes  permanencia ServicoTelefonico  \
0   Não  Mulher       0     Sim         Sim            9               Sim   
1   Não   Homem       0     Não         Não            9               Sim   
2   Sim   Homem       0     Não         Não            4               Sim   
3   Sim   Homem       1     Sim         Não           13               Sim   
4   Sim  Mulher       1     Sim         Não            3               Sim   

  MultiplasLinhas ServicoInternet SeguroOnline OnlineBackup  \
0             Não             DSL          Não          Sim   
1             Sim             DSL          Não          Não   
2             Não     FibraOptica          Não          Não   
3             Não     FibraOptica          Não          Sim   
4             Não     FibraOptica          Não          Não   

  ProtecaoDoDispositivo SuporteTecnico StreamingTV StreamingFilmes contrato  \
0                   Não            Sim         Sim             Não    UmAno   
1                   Não            Não         Não             Sim   Mensal   
2                   Sim            Não         Não             Não   Mensal   
3                   Sim            Não         Sim             Sim   Mensal   
4                   Não            Sim         Sim             Não   Mensal   

  FaturaOnline MetodoDePagamento  GastoMensal GastoTotal  
0          Sim     EnvioDeCheque         65.6      593.3  
1          Não     EnvioDeCheque         59.9      542.4  
2          Sim  ChequeEletronico         73.9     280.85  
3          Sim  ChequeEletronico         98.0    1237.85  
4          Sim     EnvioDeCheque         83.9      267.4

## Variável `GastoTotal`

A variável `GastoTotal` deveria esta no formato `float`, mas como observamos anteriormente ela está sendo classificada como `object`. Sendo assim dando uma olhada um pouco melhor na variável 

In [19]:
clientes_v05['GastoTotal'].value_counts()

           11
20.2       11
19.75       9
19.55       9
19.9        9
           ..
272         1
1426.45     1
371.6       1
6786.4      1
3707.6      1
Name: GastoTotal, Length: 6531, dtype: int64

Quando é feita uma contagem de valores que aparecem nesta variável notamos que exitem 11 valores que não foram preenchidos. Porem se olharmos melhor esses valores comparados a outras variáveis podemos ter um noção o motivo da falta destes valores.

In [20]:
clientes_v05.query('GastoTotal == " "')[['Churn', 'genero','conjuge',
                                         'dependentes','permanencia',
                                         'GastoMensal', 'GastoTotal']]

Churn  genero conjuge dependentes  permanencia  GastoMensal GastoTotal
975    Não  Mulher     Sim         Sim            0        56.05           
1775   Não  Mulher     Sim         Sim            0        20.00           
1955   Não   Homem     Não         Sim            0        61.90           
2075   Não   Homem     Sim         Sim            0        19.70           
2232   Não   Homem     Não         Sim            0        20.25           
2308   Não   Homem     Sim         Sim            0        25.35           
2930   Não  Mulher     Sim         Sim            0        73.35           
3134   Não   Homem     Sim         Sim            0        25.75           
3203   Não  Mulher     Sim         Sim            0        52.55           
4169   Não  Mulher     Sim         Sim            0        80.85           
5599   Não   Homem     Sim         Sim            0        19.85

Analisando simultaneamente as variáveis `GastoTotal` e `permanencia` é possível concluir que o  11 clientes não possuem valores de gastos, pois são clientes novos, ainda não completaram seu primeiro mês como cliente de nossa empresa. Sendo assim podemos prosseguir em trocar o tipo de valor dessa variável 

In [21]:
# criando mais uma versão do DataFrame
clientes_v06 = clientes_v05.copy()

# Realizando a conversão.
clientes_v06['GastoTotal'] = pd.to_numeric(clientes_v06['GastoTotal'], errors='coerce')

In [22]:
clientes_v06.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Churn                  7267 non-null   object 
 1   genero                 7267 non-null   object 
 2   senior                 7267 non-null   int64  
 3   conjuge                7267 non-null   object 
 4   dependentes            7267 non-null   object 
 5   permanencia            7267 non-null   int64  
 6   ServicoTelefonico      7267 non-null   object 
 7   MultiplasLinhas        7267 non-null   object 
 8   ServicoInternet        7267 non-null   object 
 9   SeguroOnline           7267 non-null   object 
 10  OnlineBackup           7267 non-null   object 
 11  ProtecaoDoDispositivo  7267 non-null   object 
 12  SuporteTecnico         7267 non-null   object 
 13  StreamingTV            7267 non-null   object 
 14  StreamingFilmes        7267 non-null   object 
 15  cont

In [23]:
clientes_v06.query('permanencia  == 0 ')

Churn  genero  senior conjuge dependentes  permanencia ServicoTelefonico  \
975    Não  Mulher       0     Sim         Sim            0               Não   
1775   Não  Mulher       0     Sim         Sim            0               Sim   
1955   Não   Homem       0     Não         Sim            0               Sim   
2075   Não   Homem       0     Sim         Sim            0               Sim   
2232   Não   Homem       0     Não         Sim            0               Sim   
2308   Não   Homem       0     Sim         Sim            0               Sim   
2930   Não  Mulher       0     Sim         Sim            0               Sim   
3134   Não   Homem       0     Sim         Sim            0               Sim   
3203   Não  Mulher       0     Sim         Sim            0               Não   
4169   Não  Mulher       0     Sim         Sim            0               Sim   
5599   Não   Homem       0     Sim         Sim            0               Sim   

     MultiplasLinhas ServicoInternet SeguroOnline OnlineBackup  \
975      SemTelefone             DSL          Sim          Sim   
1775             Não             Não  SemInternet  SemInternet   
1955             Sim             DSL          Sim          Sim   
2075             Não             Não  SemInternet  SemInternet   
2232             Não             Não  SemInternet  SemInternet   
2308             Sim             Não  SemInternet  SemInternet   
2930             Sim             DSL          Não          Sim   
3134             Sim             Não  SemInternet  SemInternet   
3203     SemTelefone             DSL          Sim          Não   
4169             Não             DSL          Sim          Sim   
5599             Não             Não  SemInternet  SemInternet   

     ProtecaoDoDispositivo SuporteTecnico  StreamingTV StreamingFilmes  \
975                    Sim            Sim          Sim             Não   
1775           SemInternet    SemInternet  SemInternet     SemInternet   
1955                   Não            Sim          Não             Não   
2075           SemInternet    SemInternet  SemInternet     SemInternet   
2232           SemInternet    SemInternet  SemInternet     SemInternet   
2308           SemInternet    SemInternet  SemInternet     SemInternet   
2930                   Sim            Sim          Sim             Não   
3134           SemInternet    SemInternet  SemInternet     SemInternet   
3203                   Sim            Sim          Sim             Não   
4169                   Sim            Não          Sim             Sim   
5599           SemInternet    SemInternet  SemInternet     SemInternet   

      contrato FaturaOnline      MetodoDePagamento  GastoMensal  GastoTotal  
975   DoisAnos          Não          CartaoCredito        56.05         NaN  
1775  DoisAnos          Não          EnvioDeCheque        20.00         NaN  
1955  DoisAnos          Sim  TransferenciaBancaria        61.90         NaN  
2075     UmAno          Sim          EnvioDeCheque        19.70         NaN  
2232  DoisAnos          Não          EnvioDeCheque        20.25         NaN  
2308  DoisAnos          Não          EnvioDeCheque        25.35         NaN  
2930  DoisAnos          Não          EnvioDeCheque        73.35         NaN  
3134  DoisAnos          Não          EnvioDeCheque        25.75         NaN  
3203  DoisAnos          Sim  TransferenciaBancaria        52.55         NaN  
4169  DoisAnos          Não          EnvioDeCheque        80.85         NaN  
5599  DoisAnos          Não          EnvioDeCheque        19.85         NaN

Agora que a variável possui uma nova tipagem, observamos que os valores que antes estavam vazios agora estão preenchidos como NaN (Not-A-Number), isso significa que não possui valores algum. Antes de solucionar esse problema é melhor verificamos se mais alguma das  variáveis do nosso DataFrame também possui valores como NaN.

In [24]:
clientes_v06.isna().sum()

Churn                     0
genero                    0
senior                    0
conjuge                   0
dependentes               0
permanencia               0
ServicoTelefonico         0
MultiplasLinhas           0
ServicoInternet           0
SeguroOnline              0
OnlineBackup              0
ProtecaoDoDispositivo     0
SuporteTecnico            0
StreamingTV               0
StreamingFilmes           0
contrato                  0
FaturaOnline              0
MetodoDePagamento         0
GastoMensal               0
GastoTotal               11
dtype: int64

Como somente a variável `GastoTotal` possui elementos vazios e sabemos agora e esses elementos são justamente aqueles que ainda não fizeram nenhum pagamento não realizando gasto com empresa preencheremos a falta destes valores com 0 (zero).

In [25]:
# Trocando os NaN por zeros.
clientes_v07 = clientes_v06.fillna(0)

In [26]:
clientes_v07.query('GastoTotal == 0')[['Churn', 'genero','conjuge',
                                         'dependentes','permanencia',
                                         'GastoMensal', 'GastoTotal']]

Churn  genero conjuge dependentes  permanencia  GastoMensal  GastoTotal
975    Não  Mulher     Sim         Sim            0        56.05         0.0
1775   Não  Mulher     Sim         Sim            0        20.00         0.0
1955   Não   Homem     Não         Sim            0        61.90         0.0
2075   Não   Homem     Sim         Sim            0        19.70         0.0
2232   Não   Homem     Não         Sim            0        20.25         0.0
2308   Não   Homem     Sim         Sim            0        25.35         0.0
2930   Não  Mulher     Sim         Sim            0        73.35         0.0
3134   Não   Homem     Sim         Sim            0        25.75         0.0
3203   Não  Mulher     Sim         Sim            0        52.55         0.0
4169   Não  Mulher     Sim         Sim            0        80.85         0.0
5599   Não   Homem     Sim         Sim            0        19.85         0.0

# Criando a variável `GastoDiario`

Criarei uma variável novo somente para ver se ela poderá acrescentar algum valor em nossos estudos.
Esta variável trará o gasto de cada cliente diariamente, posteriormente poderemos ver se ela nos ajudar ou não. 

In [28]:
clientes_v08 = clientes_v07.copy()
clientes_v08['GastoDiario'] = round(clientes_v08['GastoMensal'] /30,2) 
clientes_v08.head()

Churn  genero  senior conjuge dependentes  permanencia ServicoTelefonico  \
0   Não  Mulher       0     Sim         Sim            9               Sim   
1   Não   Homem       0     Não         Não            9               Sim   
2   Sim   Homem       0     Não         Não            4               Sim   
3   Sim   Homem       1     Sim         Não           13               Sim   
4   Sim  Mulher       1     Sim         Não            3               Sim   

  MultiplasLinhas ServicoInternet SeguroOnline OnlineBackup  \
0             Não             DSL          Não          Sim   
1             Sim             DSL          Não          Não   
2             Não     FibraOptica          Não          Não   
3             Não     FibraOptica          Não          Sim   
4             Não     FibraOptica          Não          Não   

  ProtecaoDoDispositivo SuporteTecnico StreamingTV StreamingFilmes contrato  \
0                   Não            Sim         Sim             Não    UmAno   
1                   Não            Não         Não             Sim   Mensal   
2                   Sim            Não         Não             Não   Mensal   
3                   Sim            Não         Sim             Sim   Mensal   
4                   Não            Sim         Sim             Não   Mensal   

  FaturaOnline MetodoDePagamento  GastoMensal  GastoTotal  GastoDiario  
0          Sim     EnvioDeCheque         65.6      593.30         2.19  
1          Não     EnvioDeCheque         59.9      542.40         2.00  
2          Sim  ChequeEletronico         73.9      280.85         2.46  
3          Sim  ChequeEletronico         98.0     1237.85         3.27  
4          Sim     EnvioDeCheque         83.9      267.40         2.80

# Salvado o DataFrame

Agora que o DataFrame esta com seus dados limpos e ajustados, podemos passar para a próxima etapa, a análise dos dados. 

In [30]:
clientes_v08.to_json('dados/Telco-Customer-Churn-Limpo.json')